feat: add `@stdlib/blas/base/srotg`

Pranavchiku opened the pull request:
Fixes #280.

## Checklist

- [x] Add readme.md
- [x] repl.txt
- [x] types
- [x] package.json
- [x] lib
- [x] examples
- [x] benchmark
- [x] test 
* * *

@kgryte

Pranavchiku committed:
Add readme

Pranavchiku committed:
Add main.js, assign.js, index.js in lib/

Pranavchiku committed:
add package.json

Pranavchiku committed:
add examples

Pranavchiku committed:
add js example in readme

Pranavchiku committed:
fix indentation in package.json

Pranavchiku committed:
add benchmark for js

Pranavchiku committed:
add repl.txt

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,91 @@
+		if ( abs( a ) > abs( b ) ) {
+			z = s;
+		}
+		if ( abs( b ) >= abs( a ) && c !== 0.0 ) {
```
I think you want `else if` here.

Pranavchiku replied:
https://github.com/xianyi/OpenBLAS/blob/develop/lapack-netlib/BLAS/SRC/srotg.f#L104 it had both IF statements, so...

Pranavchiku replied:
Both will be equivalent imo, will change to `else if`

kgryte replied:
Actually, we should move L75 to L72 and do `else if ( aa > ab ) { z = s; }`. The second condition is more specific than the second.

kgryte replied:
Correct. We're wasting cycles without the `else if`. As right now, L75 is always exercised atm even when L72 is true.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,91 @@
+	var s;
+	var z;
+
+	roe = b;
```
What does `roe` stand for?

Pranavchiku replied:
Not sure, no such description is provided.

kgryte replied:
`roe` appears to be used exclusively for computing the sign. In which case, I'd rename this variable to `sign`, and I'd refactor as

```
aa = abs( a );
ab = abs( b );
if ( aa > ab ) {
    sign = copysign( 1.0, a );
} else {
    sign = copysign( 1.0, b );
}
```

Pranavchiku replied:
Why use `copysign`? We are doing it at `r *= copysign( 1.0, roe )`

Pranavchiku replied:
```
aa = abs( a );
ab = abs( b );
sign = b;
if ( aa > ab ) {
    sign = a;
}
```

kgryte replied:
You only need a single assignment and temporary variable. Compare

```
roe = 2;
x = copysign( 1.0, roe ) * r;
```

which involves two temp variables, one explicit and the other implicit.

While

```
roe = copysign( 1.0, a );
x = sign * r;
```

only one. In short, why is the assignment of `roe` to either `a` or `b` actually necessary?

Pranavchiku replied:
Okay, got it, in implementation we just change `r *= sign`.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,91 @@
+	if ( abs( a ) > abs( b ) ) {
+		roe = a;
+	}
+	scale = abs( a ) + abs( b );
```
You're computing `abs(a)` and `abs(b)` multiple times. Prefer caching the computed values.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,91 @@
+* var out = srotg( 0.0, 2.0, new Float32Array( 4 ), 1, 0 );
+*  // returns <Float32Array>[ 2.0, 1.0, 0.0, 1.0 ]
+*/
+function srotg( a, b, out, stride, offset ) {
```
This implementation follows the OpenBLAS reference implementation, correct?

Pranavchiku replied:
Yes, this one https://github.com/xianyi/OpenBLAS/blob/develop/lapack-netlib/BLAS/SRC/srotg.f

kgryte replied:
Confirmed: https://github.com/xianyi/OpenBLAS/blob/6c1076e133cb034a7be3f66934f88ed37319add9/lapack-netlib/BLAS/SRC/srotg.f

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,91 @@
+		z = 0.0;
+	} else {
+		r = scale * sqrt( pow( a/scale, 2 ) + pow( b/scale, 2 ) );
+		r = copysign( 1.0, roe ) * r;
```
You can use in-place assignment.

```
r *= copysign( ... );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/package.json:
```diff
@@ -0,0 +1,81 @@
+{
+  "name": "@stdlib/blas/base/srotg",
+  "version": "0.0.0",
+  "description": "Constructs a Givens rotation from two values `a` and `b`.",
```
```suggestion
  "description": "Construct a Givens plane rotation.",
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/package.json:
```diff
@@ -0,0 +1,81 @@
+    "blas",
+    "level 1",
+    "srotg",
+    "dot",
+    "product",
+    "dot product",
+    "scalar",
+    "scalar product",
+    "inner",
+    "inner product",
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/index.js:
```diff
@@ -0,0 +1,64 @@
+*
+* @module @stdlib/blas/base/srotg
+*
+*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/index.js:
```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+/**
+* Construct a Givens rotation from two values `a` and `b`.
```
```suggestion
* Construct a Givens plane rotation.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/index.js:
```diff
@@ -0,0 +1,64 @@
+
+// EXPORTS //
+
+module.exports = main;
```
Check other packages, but we've started adding a comment at the bottom of `index.js` files for those APIs having additional methods.

Pranavchiku committed:
Applied suggestion from code review in batch

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,129 @@
+
+# srotg
+
+> Constructs a Given plane rotation for single-precision values `a` and `b`.
```
```suggestion
> Constructs a Givens plane rotation for single-precision values `a` and `b`.
```

Should be updated here and below.

Pranavchiku committed:
refactor variable roe

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,95 @@
+		z = 1.0;
+		if ( aa > ab ) {
+			z = s;
+		} else if ( c !== 0.0 ) {
```
Nice! Good catch to eliminate the redundant `aa <= ab` check.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,95 @@
+*
+* @example
+* var Float32Array = require( '@stdlib/array/float32' );
+* var out = srotg( 0.0, 2.0, new Float32Array( 4 ), 1, 0 );
```
```suggestion
*
* var out = srotg( 0.0, 2.0, new Float32Array( 4 ), 1, 0 );
```

Pranavchiku committed:
update description in readme

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,95 @@
+/**
+* Construct a Givens plane rotation.
+*
+* @private
```
```suggestion
```

This API is part of the public package interface, so we can mark as public.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,95 @@
+// MAIN //
+
+/**
+* Construct a Givens plane rotation.
```
```suggestion
* Constructs a Givens plane rotation.
```

For the actual function, we use the active voice. E.g., `Constructs`, rather than `Construct`. For the module and package description, we use the imperative voice. E.g., `Construct`.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/repl.txt:
```diff
@@ -0,0 +1,26 @@
+    > var out = {{alias}}( 0.0, 2.0 )
+    <Float32Array>[ 2.0, 1.0, 0.0, 1.0 ]
+
+    See Also
```
We should also document the `assign` API, as it is part of the public interface.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/benchmark/benchmark.js:
```diff
@@ -0,0 +1,53 @@
+	}
+	b.pass( 'benchmark finished' );
+	b.end();
+});
```
We should also benchmark the assign API.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,129 @@
+-   **a**: rotational elimination parameter.
+-   **b**: rotational elimination parameter.
+
+</section>
```
We should also document the `assign` API as it is part of the public interface.

Pranavchiku committed:
document assign api in readme

Pranavchiku committed:
document assign api usage in repl.txt

Pranavchiku committed:
apply suggestion from code review in batch

Pranavchiku committed:
add tests

Pranavchiku committed:
Add index.d.ts

Pranavchiku committed:
add test.ts in types/

Pranavchiku committed:
Add NaN checks

Pranavchiku committed:
add new tests

Pranavchiku commented:
I created a few more test, kindly have a look and do suggest if stronger tests are required.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:
```diff
@@ -0,0 +1,137 @@
+	t.end();
+});
+
+tape( 'the function throws NaN when a is not number', function test( t ) {
```
Instead of `throws`, I believe you mean `returns`.

Pranavchiku replied:
Yeah. Will change.

kgryte commented:
For tests, I'd highly recommend including all the test cases found in the OpenBLAS test suite. Our implementation should return the same results.

Pranavchiku commented:
Could you please provide the link for the test suite?

kgryte commented:
https://github.com/xianyi/OpenBLAS/blob/develop/lapack-netlib/BLAS/TESTING/sblat1.f

Pranavchiku commented:
Some tests fail with 
```console
not ok 8 returns expected value
  ---
    operator: deepEqual
    expected: |-
      { 0: -0.5, 1: 1.659999966621399, 2: 0.6000000238418579, 3: 0.800000011920929 }
    actual: |-
      { 0: -0.5, 1: 1.6666666269302368, 2: 0.6000000238418579, 3: 0.800000011920929 }
    at: Test.test (/home/pranavchiku/stdlib/lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:68:5)
  ...
```
As of now, I have used a workaround keeping expected[1] = 1.6666 and not 1.66.

Pranavchiku committed:
add new tests

Pranavchiku commented:
@kgryte kindly have a look and do tell any changes need to be made.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
````diff
@@ -0,0 +1,145 @@
+```javascript
+var Float32Array = require( '@stdlib/array/float32' );
+
+var a = 0.0;
+var b = 2.0;
+
+var out = srotg( a, b );
````
```suggestion
var out = srotg( 0.0, 2.0 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
````diff
@@ -0,0 +1,145 @@
+Constructs a Givens plane rotation for single-precision values `a` and `b`.
+
+```javascript
+var Float32Array = require( '@stdlib/array/float32' );
+
````
```suggestion
```

This import is not used.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,145 @@
+
+#### srotg.assign( a, b, out, stride, offset )
+
+Constructs Givens plane rotations for single-precision floating-point values and assigns results to output array.
```
```suggestion
Constructs a Givens plane rotation for single-precision floating-point values and assigns results to an output array.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,145 @@
+
+The function has the following parameters:
+
+-   **a**: rotational elimination parameter.
+-   **b**: rotational elimination parameter.
```
These both have the same description?

Pranavchiku replied:
Yeah.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,145 @@
+var i;
+
+for ( i = 0; i < 100; i++ ) {
+    a = round( randu()*10.0 ) - 5.0;
```
Instead of `round(randu())`, let's use `discreteUniform( -5, 5 )`.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,145 @@
+
+<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->
+
+<section class="related">
+
+* * *
+
+## See Also
+
+-   <span class="package-name">[`@stdlib/blas/base/daxpy`][@stdlib/blas/base/daxpy]</span><span class="delimiter">: </span><span class="description">multiply a vector x by a constant and add the result to y.</span>
+-   <span class="package-name">[`@stdlib/blas/base/dscal`][@stdlib/blas/base/dscal]</span><span class="delimiter">: </span><span class="description">multiply a double-precision floating-point vector by a constant.</span>
+-   <span class="package-name">[`@stdlib/blas/base/gscal`][@stdlib/blas/base/gscal]</span><span class="delimiter">: </span><span class="description">multiply a vector by a constant.</span>
+-   <span class="package-name">[`@stdlib/blas/base/saxpy`][@stdlib/blas/base/saxpy]</span><span class="delimiter">: </span><span class="description">multiply a vector x by a constant and add the result to y.</span>
+
```
```suggestion
<section class="related">

```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,145 @@
+[blas]: http://www.netlib.org/blas
+
+[srotg]: http://www.netlib.org/lapack/explore-html/df/d28/group__single__blas__level1.html
+
+<!-- <related-links> -->
+
+[@stdlib/blas/base/daxpy]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/blas/base/daxpy
+
+[@stdlib/blas/base/dscal]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/blas/base/dscal
+
+[@stdlib/blas/base/gscal]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/blas/base/gscal
+
+[@stdlib/blas/base/saxpy]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/blas/base/saxpy
+
+<!-- </related-links> -->
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/benchmark/benchmark.js:
```diff
@@ -0,0 +1,80 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = ( randu()*10.0 ) - 5.0;
+		y = ( randu()*10.0 ) - 5.0;
```
Instead of generating new random numbers for each iteration, let's sample from a list of pre-generated values (e.g., `x[ i%x.length ]`. Otherwise, we run the risk of benchmarking `randu`, rather than the main function.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/benchmark/benchmark.js:
```diff
@@ -0,0 +1,80 @@
+		}
+	}
+	b.toc();
+	if ( z !== out ) {
```
I'd used the L49 check here, as well, which is a bit more robust to potential future compiler optimizations.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/benchmark/benchmark.js:
```diff
@@ -0,0 +1,80 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = ( randu()*10.0 ) - 5.0;
```
Same comment as above.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/repl.txt:
```diff
@@ -0,0 +1,62 @@
+
+    Examples
+    --------
+    // Standard usage:
+    > var out = {{alias}}( 0.0, 2.0 )
```
```suggestion
    > var out = {{alias}}( 0.0, 2.0 )
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/repl.txt:
```diff
@@ -0,0 +1,62 @@
+
+{{alias}}.assign( a, b, out, stride, offset )
+    Constructs a Givens rotation from two single-precision 
+    floating-point numbers and assigns the results to a given
+    output array.
```
```suggestion
    floating-point numbers and assigns the results to an output
    array.
```

You may need to rewrap.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Inteface describing `srotg`.
+ */
+interface Srotg {
```
Let's rename `Srotg` to `Routine`. The `Srotg` capitalization looks a bit funny given that the rest of the name is also abbreviation and/or an acronym.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+	* var out = srotg( 6.0, -8.0 );;
+	* // returns [ 10.0, -1.666, -0.6, 0.8 ]
+	*/
+	( a: number, b: number ): Array<number>;
```
We return a `Float32Array`, correct?

Pranavchiku replied:
Yeah.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+    /**
+    * Constructs a Givens plane rotation.
+    *
+    * @param {number} a - first value
```
In TypeScript files, we don't include `{<type>}` annotations as the dtypes are present in the signatures.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+    * var bool = ( y === out );
+    * // returns true
+    */
+	assign( a: number, b: number, out: Collection, stride: number, offset: number ): Collection; // tslint-disable-line max-line-length
```
s/Collection/Float32Array/. This is a typed interface.

Pranavchiku replied:
I am not able to understand the suggestion here.

Pranavchiku replied:
Do you mean `out: Collection|Float32Array`?

kgryte replied:
s// is `sed` syntax, meaning substitute the latter for the former. In this case, this BLAS interface is specifically intended for single-precision floating-point operands. In which case, `out` should be limited to `Float32Array`, rather than the more general `Collection`. In particular, for the native interface, we cannot readily pass a generic array-like object down to C. Instead, we'll want to pass down only `Float32Array` instances.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+* // returns [ 2.0, 1.0, 0.0, 1.0 ]
+*
+* @example
+* var out = srotg( 6.0, -8.0 );;
```
```suggestion
* var out = srotg( 6.0, -8.0 );
```

kgryte replied:
Instead of another two-param example, I'd use an `.assign` example that way we have one example for each API.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+
+// The function returns a collection...
+{
+	srotg( 0.0, 2.0 ); // $ExpectType Collection
```
Will need to be updated to reflect return type.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+
+// The compiler throws an error if the function is provided an argument which is not a number...
+{
+	srotg( true, false ); // $ExpectError
```
Typically, convention is to hold one param constant while supplying wrong values for the other param and then repeating for the other param, where the first param is constant (and valid) and the second param is supplied various invalid arg types.

kgryte replied:
This is still unresolved. Are you planning on updating?

Pranavchiku replied:
Yeah I will do that.

kgryte replied:
@Pranavchiku This still needs to be updated.

Pranavchiku replied:
Updated this.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+	srotg( {}, {} ); // $ExpectError
+}
+
+// The compiler throws an error if the function is provided insufficient arguments...
```
```suggestion
// The compiler throws an error if the function is provided an unsupported number of arguments...
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+	srotg( 1.0, 2.0, 3.0 ); // $ExpectError
+}
+
+// Attached to the main export is an `assign` method which returns an array-like object containing numbers...
```
Should be changed once output type updated.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+{
+	const out = [ 0.0, 0.0, 0.0, 0.0 ];
+
+	srotg.assign( true, true, out, 1, 0 ); // $ExpectError
```
Same comment as above.

kgryte replied:
If the above tests are updated, these should be updated, as well.

kgryte replied:
@Pranavchiku This still needs to be updated.

Pranavchiku replied:
Updated.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+
+// Attached to the main export is an `assign` method which returns an array-like object containing numbers...
+{
+	const out = [ 0.0, 0.0, 0.0, 0.0 ];
```
Will need to be updated to `Float32Array` here and elsewhere.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/examples/index.js:
```diff
@@ -0,0 +1,35 @@
+var i;
+
+for ( i = 0; i < 100; i++ ) {
+	a = round( randu()*10.0 ) - 5.0;
```
Use discrete uniform PRNG.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,95 @@
+* @param {number} offset - starting index
+* @returns {Float32Array} destination array
+*
+*
```
```suggestion
```

Only one blank line separating param list from examples.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,95 @@
+* @param {number} a - first value
+* @param {number} b - second value
+* @param {Float32Array} out - destination array
+* @param {number} stride - index increment
+* @param {number} offset - starting index
```
`integer` and `NonNegativeInteger`, respectively.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/index.js:
```diff
@@ -0,0 +1,63 @@
+* out = srotg( 6.0, -8.0 );
+* // returns <Float32Array>[ 10.0, -1.666, -0.6, 0.8 ]
+*
+*
```
```suggestion
```

One blank line separating examples.

@Planeshifter We should create a lint rule for this.

Pranavchiku committed:
apply suggestion from code review in readme

Pranavchiku committed:
update index.js in examples/

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/package.json:
```diff
@@ -0,0 +1,74 @@
+  ],
+  "main": "./lib",
+  "browser": "./lib/main.js",
+  "gypfile": true,
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.assign.js:
```diff
@@ -0,0 +1,130 @@
+	for ( i = 0; i < values.length; i++ ) {
+		expected[i] = new Float32Array( expected[i] );
+		out = new Float32Array( 4 );
+		srotg.assign( values[i][0], values[i][1], out, 1, 0 );
```
You're passing in doubles. You need to wrap with `float64ToFloat32(...)` in order to "cast" a double-precision scalar to a single-precision equiv. In short, `float64ToFloat32( 0.3 ) !== 0.3`.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.assign.js:
```diff
@@ -0,0 +1,130 @@
+		srotg.assign( values[i][0], values[i][1], out, 1, 0 );
+		t.deepEqual( out, expected[i], 'returns expected value' );
+	}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.assign.js:
```diff
@@ -0,0 +1,130 @@
+	t.end();
+});
+
+tape( 'the function computes a Givens plane rotation ( a = NaN, b = NaN )', function test(t) {
```
```suggestion
tape( 'the function computes a Givens plane rotation (a = NaN, b = NaN)', function test(t) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.assign.js:
```diff
@@ -0,0 +1,130 @@
+	for ( i = 0; i < actual.length; i++ ) {
+		t.strictEqual( isnan( actual[i] ), true, 'returns expected value' );
+	}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.assign.js:
```diff
@@ -0,0 +1,130 @@
+});
+
+tape( 'the function returns NaN when a is not number', function test( t ) {
+	var actual;
+	var values = [
+		'a',
+		undefined,
+		NaN,
+		{},
+		function foo() {}
+	];
+	var i;
+	var j;
```
In general, unless only a single variable is declared, convention is not to mix declaration and assignment. In which case, the assignment to `values` should happen separately before the `for` loop.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.assign.js:
```diff
@@ -0,0 +1,130 @@
+	t.end();
+});
+
+tape( 'the function returns NaN when a is not number', function test( t ) {
```
These tests are not necessary. Base BLAS implementations assume that they're provided numeric/valid values.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.assign.js:
```diff
@@ -0,0 +1,130 @@
+	t.end();
+});
+
+tape( 'the function returns NaN when b is not number', function test( t ) {
```
These tests are not needed.

Pranavchiku committed:
update benchmark.js

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.js:
```diff
@@ -0,0 +1,40 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
Copyright year.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:
```diff
@@ -0,0 +1,153 @@
+
+	for ( i = 0; i < values.length; i++ ) {
+		expected[i] = new Float32Array( expected[i] );
+		out = srotg( values[i][0], values[i][1] );
```
Same comment as for `test.assign.js`.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:
```diff
@@ -0,0 +1,153 @@
+});
+
+tape( 'the function computes a Givens plane rotation', function test( t ) {
+	var expected = [
```
Same comment as in `test.assign.js` regarding declaration and assignment.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:
```diff
@@ -0,0 +1,153 @@
+	actual = srotg( 0.0, 0.0 );
+
+	t.deepEqual( actual, expected, 'returns expected value' );
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:
```diff
@@ -0,0 +1,153 @@
+	actual = srotg( 1.0, 0.0 );
+
+	t.deepEqual( actual, expected, 'returns expected value' );
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:
```diff
@@ -0,0 +1,153 @@
+	t.end();
+});
+
+tape( 'the function computes a Givens plane rotation ( a = NaN, b = NaN )', function test(t) {
```
```suggestion
tape( 'the function computes a Givens plane rotation (a = NaN, b = NaN)', function test(t) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:
```diff
@@ -0,0 +1,153 @@
+	for ( i = 0; i < actual.length; i++ ) {
+		t.strictEqual( isnan( actual[i] ), true, 'returns expected value' );
+	}
+
```
```suggestion
```

Pranavchiku committed:
update repl.txt

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:
```diff
@@ -0,0 +1,153 @@
+	t.end();
+});
+
+tape( 'the function returns NaN when a is not number', function test( t ) {
```
These and the following tests are not needed and should be removed.

kgryte reviewed (changes requested):
@Pranavchiku Thanks for continuing to work on this. Failing tests may be due to a lack of double-precision to single-precision casting.

Pranavchiku committed:
update assign.js

Pranavchiku committed:
update index.js

Pranavchiku committed:
update package.json

Pranavchiku committed:
update test.assign.js

Pranavchiku committed:
udpate test.main.js

Pranavchiku committed:
udpate test.main.js

Pranavchiku committed:
update index.d.ts

Pranavchiku committed:
update test.ts and index.d.ts

Pranavchiku committed:
update test.ts

Pranavchiku commented:
I applied suggestions from code review and left comments on a few of them which I was not able to understand clearly, those are majorly from `docs/types/*`.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,118 @@
+
+#### srotg.assign( a, b, out, stride, offset )
+
+Constructs a Givens plane rotation for single-precision floating-point values and assigns results to an output array.
```
```suggestion
Constructs a Givens plane rotation for single-precision floating-point values `a` and `b` and assigns results to an output array.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/benchmark/benchmark.js:
```diff
@@ -0,0 +1,77 @@
+
+bench( pkg, function benchmark( b ) {
+	var out;
+	var x = filledarrayBy( 100, 'float64', discreteUniform( -5, 5 ) );
```
We have new `@stdlib/random/array` packages which more succinctly achieve this. I'll update on my end.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+
+// TypeScript Version: 2.0
+
+/// <reference types="@stdlib/types"/>
+
+import { Collection } from '@stdlib/types/object';
+
```
```suggestion
```

No longer used.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+	/**
+	* Constructs a Givens plane rotation.
+	*
+	* @param a - first value
+    * @param b - second value
```
```suggestion
	* @param a - rotational elimination parameter
    * @param b - rotational elimination parameter
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+	*
+	* @example
+	* var out = srotg( 6.0, -8.0 );;
+	* // returns [ 10.0, -1.666, -0.6, 0.8 ]
```
Is `-1.666` exact or approximate? I.e., should we preface with a `~`?

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+    /**
+    * Constructs a Givens plane rotation.
+    *
+    * @param a - first value
+    * @param b - second value
```
```suggestion
    * @param a - rotational elimination parameter
    * @param b - rotational elimination parameter
```

kgryte replied:
@Pranavchiku We need to be careful about force pushing (see <https://github.com/stdlib-js/stdlib/compare/9051e58ce2cae71e0f9325532576954d88556e87..10c5ea3b27ace0868462791da1d19122899f6f7f>). In this case, force pushing reverted committed changes. It is not clear why force pushing should be necessary when updating the PR, rather than merging and dealing with any merge conflicts locally.

I'd suggest going back through the force push comparison and ensuring that all intentional changes that were reverted by the force push are restored. Once done, LMK, and I can do another review.

Pranavchiku replied:
Ahh, looks like I forgot to pull up the changes and then forced pushed ( thinking I reverted some changes ). I will iterate over as soon as possible.

Pranavchiku replied:
resolved.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+/**
+* Constructs a Givens plane rotation.
+*
+* @param a - first value
+* @param b - second value
```
```suggestion
* @param a - rotational elimination parameter
* @param b - rotational elimination parameter
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+* // returns [ 2.0, 1.0, 0.0, 1.0 ]
+*
+* @example
+* var y = srotg.assign( 0.0, 2.0, out, 1, 0 );
```
You need to define `out` here. Currently, this is an undefined variable. Should be a `Float32Array`.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+	srotg( {}, {} ); // $ExpectError
+}
+
+// The compiler throws an error if the function is provided an unsupported numbder of arguments...
```
```suggestion
// The compiler throws an error if the function is provided an unsupported number of arguments...
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+	srotg( 1.0, 2.0, 3.0 ); // $ExpectError
+}
+
+// Attached to the main export is an `assign` method which returns a Float32Array containing numbers...
```
```suggestion
// Attached to the main export is an `assign` method which returns a Float32Array...
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+	srotg.assign( {}, {}, out, 1, 0 ); // $ExpectError
+}
+
+// The compiler throws an error if the `assign` method is provided a third argument which is not an array-like object...
```
```suggestion
// The compiler throws an error if the `assign` method is provided a third argument which is not a Float32Array...
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,94 @@
+/**
+* Constructs a Givens plane rotation.
+*
+* @param {number} a - first value
+* @param {number} b - second value
```
```suggestion
* @param {number} a - rotational elimination parameter
* @param {number} b - rotational elimination parameter
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/main.js:
```diff
@@ -0,0 +1,48 @@
+/**
+* Construct a Givens plane rotation.
+*
+* @param {number} a - first value
+* @param {number} b - second value
```
```suggestion
* @param {number} a - rotational elimination parameter
* @param {number} b - rotational elimination parameter
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/package.json:
```diff
@@ -0,0 +1,73 @@
+    }
+  ],
+  "main": "./lib",
+  "browser": "./lib/main.js",
```
```suggestion
```

Currently, this would break browser bundles, as the `assign` method would be excluded from the main export. This is found in other BLAS packages as `index.js` chooses between a JS vs native implementation. In the current implementation, we don't have that.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.assign.js:
```diff
@@ -0,0 +1,88 @@
+		t.strictEqual( isnan( actual[i] ), true, 'returns expected value' );
+	}
+	t.end();
+});
```
Ideally, we'd also include tests for stride and offset behavior for the output array. See other strided array implementations for how this is done.

Pranavchiku replied:
So we are expecting the length of `out` to be greater than 4 and the stride can have any value?

Pranavchiku replied:
Why do we need these kind of tests here, when we are sure about returning 4 values ( in form of Float32Array ) with two double inputs?

kgryte reviewed (changes requested):
Mostly good. An example needs to be fixed. Some tests in `test.assign.js` should be added. And two comments regarding TypeScript declaration tests.

kgryte commented:
Also, we have some lint errors in the TypeScript files: https://github.com/stdlib-js/stdlib/actions/runs/4168587179/jobs/7215562644.

Pranavchiku committed:
resolve lint errors

Pranavchiku committed:
declare variable out in index.d.ts

Pranavchiku committed:
preface ~ in index.d.ts

Pranavchiku committed:
add test for both stride and offset

Pranavchiku committed:
fix indentation spacing in index.d.ts

Pranavchiku commented:
I have added simple tests for both positive and negative stride, and for offset as well kindly have a look and do tell me if more tests need to be added. 
All tests and benchmarks cleared, kindly have a look at this PR. @kgryte

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,118 @@
+
+# srotg
+
+> Constructs a Givens plane rotation for single-precision values `a` and `b`.
```
```suggestion
> Construct a Givens plane rotation for single-precision values `a` and `b`.
```

kgryte committed:
docs: fix description

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/benchmark/benchmark.js:
```diff
@@ -0,0 +1,77 @@
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		z = srotg.assign( x[ i%x.length ], y[ i%y.length ], out, 1, 0 );
+		if ( typeof out !=='object' ) {
```
```suggestion
		if ( typeof z !=='object' ) {
```

kgryte committed:
bench: change assertion to avoid potential compiler optimizations

We want to check the return value instead of `out`, as otherwise a compiler might move the check out of the loop.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/benchmark/benchmark.js:
```diff
@@ -0,0 +1,77 @@
+var bench = require( '@stdlib/bench' );
+var discreteUniform = require( '@stdlib/random/base/discrete-uniform' ).factory;
+var filledarrayBy = require( '@stdlib/array/filled-by' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
```
```suggestion
var isnanf = require( '@stdlib/math/base/assert/is-nanf' );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/benchmark/benchmark.js:
```diff
@@ -0,0 +1,77 @@
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		out = srotg( x[ i%x.length ], y[ i%y.length ] );
+		if ( isnan( out[ i%4 ] ) ) {
```
```suggestion
		if ( isnanf( out[ i%4 ] ) ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/benchmark/benchmark.js:
```diff
@@ -0,0 +1,77 @@
+		}
+	}
+	b.toc();
+	if ( isnan( out[ i%4 ] ) ) {
```
```suggestion
	if ( isnanf( out[ i%4 ] ) ) {
```

kgryte committed:
bench: use single-precision package when testing single-precision values

Pranavchiku committed:
add changes removed by force-push

Pranavchiku commented:
Iterated over suggested changes and added missing suggestions. This PR can have a final review. Apologies for inconvenience caused.

Pranavchiku committed:
fix error in benchmark.js

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,118 @@
+
+# srotg
+
+> Construct a Givens plane rotation for single-precision values `a` and `b`.
```
```suggestion
> Construct a Givens plane rotation.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,118 @@
+
+#### srotg( a, b )
+
+Constructs a Givens plane rotation for single-precision values `a` and `b`.
```
```suggestion
Constructs a Givens plane rotation for single-precision floating-point values `a` and `b`.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/README.md:
```diff
@@ -0,0 +1,118 @@
+
+#### srotg.assign( a, b, out, stride, offset )
+
+Constructs a Givens plane rotation for single-precision floating-point values and assigns results to an output array.
```
```suggestion
Constructs a Givens plane rotation for single-precision floating-point values `a` and `b` and assigns results to an output array.
```

kgryte committed:
Apply suggestions from code review

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+
+// TypeScript Version: 2.0
+
+/// <reference types="@stdlib/types"/>
+
```
```suggestion
```

This isn't necessary as we are not importing any `stdlib` types.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+    *
+    * @example
+    * var Float32Array = require( '@stdlib/array/float32' );
+    * var srotg = require( '@stdlib/blas/base/srotg' );
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+    * @returns destination array
+    *
+    * @example
+    * var Float32Array = require( '@stdlib/array/float32' );
```
```suggestion
    * var Float32Array = require( `@stdlib/array/float32` );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+    *
+    * @example
+    * var out = srotg( 0.0, 2.0 );
+    * // returns [ 2.0, 1.0, 0.0, 1.0 ]
```
```suggestion
    * // returns <Float32Array>[ 2.0, 1.0, 0.0, 1.0 ]
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+    *
+    * @example
+    * var out = srotg( 6.0, -8.0 );
+    * // returns [ 10.0, ~-1.666, -0.6, 0.8 ]
```
```suggestion
    * // returns <Float32Array>[ 10.0, ~-1.666, -0.6, 0.8 ]
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+*
+* @example
+* var out = srotg( 0.0, 2.0 );
+* // returns [ 2.0, 1.0, 0.0, 1.0 ]
```
```suggestion
* // returns <Float32Array>[ 2.0, 1.0, 0.0, 1.0 ]
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,91 @@
+*
+* var out = new Float32Array( 4 );
+* var y = srotg.assign( 0.0, 2.0, out, 1, 0 );
+* // returns [ 2.0, 1.0, 0.0, 1.0 ]
```
```suggestion
* // returns <Float32Array>[ 2.0, 1.0, 0.0, 1.0 ]
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+
+// TESTS //
+
+// The function returns a collection...
```
```suggestion
// The function returns a Float32Array...
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+
+// The compiler throws an error if the `assign` method is provided a fourth argument which is not a number...
+{
+	const out = [ 0.0, 0.0, 0.0, 0.0 ];
```
```suggestion
	const out = new Float32Array( 4 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+
+// The compiler throws an error if the `assign` method is provided a fifth argument which is not a number...
+{
+	const out = [ 0.0, 0.0, 0.0, 0.0 ];
```
```suggestion
	const out = new Float32Array( 4 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+
+// The compiler throws an error if the `assign` method is provided an unsupported number of arguments...
+{
+	const out = [ 0.0, 0.0, 0.0, 0.0 ];
```
```suggestion
	const out = new Float32Array( 4 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/main.js:
```diff
@@ -0,0 +1,48 @@
+// MAIN //
+
+/**
+* Construct a Givens plane rotation.
```
```suggestion
* Constructs a Givens plane rotation.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/main.js:
```diff
@@ -0,0 +1,48 @@
+*
+* @param {number} a - rotational elimination parameter
+* @param {number} b - rotational elimination parameter
+* @returns {Float32Array} destination array
```
```suggestion
* @returns {Float32Array} output array
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/package.json:
```diff
@@ -0,0 +1,72 @@
+    "benchmark": "./benchmark",
+    "doc": "./docs",
+    "example": "./examples",
+    "include": "./include",
+    "lib": "./lib",
+    "src": "./src",
```
```suggestion
    "lib": "./lib",
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.assign.js:
```diff
@@ -0,0 +1,144 @@
+var tape = require( 'tape' );
+var Float32Array = require( '@stdlib/array/float32' );
+var float64ToFloat32 = require( '@stdlib/number/float64/base/to-float32' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
```
This should be `isnanf`.

kgryte replied:
```suggestion
var isnanf = require( '@stdlib/math/base/assert/is-nanf' );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:
```diff
@@ -0,0 +1,109 @@
+var tape = require( 'tape' );
+var Float32Array = require( '@stdlib/array/float32' );
+var float64ToFloat32 = require( '@stdlib/number/float64/base/to-float32' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
```
```suggestion
var isnanf = require( '@stdlib/math/base/assert/is-nanf' );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.main.js:
```diff
@@ -0,0 +1,109 @@
+
+	actual = srotg( NaN, NaN );
+	for ( i = 0; i < actual.length; i++ ) {
+		t.strictEqual( isnan( actual[i] ), true, 'returns expected value' );
```
```suggestion
		t.strictEqual( isnanf( actual[i] ), true, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/test/test.assign.js:
```diff
@@ -0,0 +1,144 @@
+
+	actual = srotg( NaN, NaN, new Float32Array( 4 ), 1, 0 );
+	for ( i = 0; i < actual.length; i++ ) {
+		t.strictEqual( isnan( actual[i] ), true, 'returns expected value' );
```
```suggestion
		t.strictEqual( isnanf( actual[i] ), true, 'returns expected value' );
```

kgryte committed:
Apply suggestions from code review

kgryte reviewed (changes requested):
@Pranavchiku Based on most recent changes, the only thing left on this PR are the updates to the TypeScript declaration tests. Once the requested changes are resolved, this should be ready for merge.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,88 @@
+// TypeScript Version: 2.0
+
+/**
+ * Inteface describing `srotg`.
+ */
```
```suggestion
* Inteface describing `srotg`.
*/
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,88 @@
+ * Inteface describing `srotg`.
+ */
+interface Routine {
+    /**
```
@Pranavchiku This all needs to be converted to TAB indentation.

Pranavchiku replied:
I guess I have provided TAB indentation. I checked it manually as well, I am not sure why it is throwing lint errors.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,88 @@
+    * // returns <Float32Array>[ 10.0, ~-1.666, -0.6, 0.8 ]
+    */
+    ( a: number, b: number ): Float32Array; 
+    /**
```
```suggestion

    /**
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/index.d.ts:
```diff
@@ -0,0 +1,88 @@
+    * var out = srotg( 6.0, -8.0 );
+    * // returns <Float32Array>[ 10.0, ~-1.666, -0.6, 0.8 ]
+    */
+    ( a: number, b: number ): Float32Array; 
```
```suggestion
    ( a: number, b: number ): Float32Array;
```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,112 @@
+}
+
+// Attached to the main export is an `assign` method which returns a Float32Array...
+{	
```
```suggestion
{
```

kgryte committed:
Apply suggestions from code review

Pranavchiku commented:
Aah, sorry for that, I will do that soon and ping you once done.

Pranavchiku committed:
provide tab indent in index.d.ts

Pranavchiku commented:
I do not know why typescript linters are not working for me, that is why facing so much trouble, once this PR gets merged I will again install node-modules.

Pranavchiku committed:
add trailing whitespace in index.d.ts

Pranavchiku committed:
remove trailing white space

Pranavchiku commented:
Finally got all tests passed! @kgryte you can have a final review, apologies for highly messed up PR!

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,94 @@
+	} else {
+		sign = copysign( 1.0, b );
+	}
+	scale = aa + ab;
```
Just realized this, but we should be emulating single-precision floating-point arithmetic here. E.g.,

```javascript
scale = float64ToFloat32( aa + ab );
```

This applies here and the result of each operation.

kgryte replied:
Currently, operations are being performed in "extended" precision. In order to match C, we need to "convert" the result of each operation to a `float32`.

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,94 @@
+
+// MODULES //
+
+var abs = require( '@stdlib/math/base/special/abs' );
```
```suggestion
var absf = require( '@stdlib/math/base/special/absf' );
```

This should be updated below, as well. `s/abs/absf/`

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,94 @@
+// MODULES //
+
+var abs = require( '@stdlib/math/base/special/abs' );
+var sqrt = require( '@stdlib/math/base/special/sqrt' );
```
s/sqrt/sqrtf/

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,94 @@
+var abs = require( '@stdlib/math/base/special/abs' );
+var sqrt = require( '@stdlib/math/base/special/sqrt' );
+var pow = require( '@stdlib/math/base/special/pow' );
+var copysign = require( '@stdlib/math/base/special/copysign' );
```
s/copysign/copysignf/

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,94 @@
+		z = 0.0;
+	} else {
+		r = scale * sqrt( pow( a/scale, 2 ) + pow( b/scale, 2 ) );
+		r *= sign;
```
For `float32` emulation, this will become

```javascript
r = float64ToFloat32( r * sign );
```

Pranavchiku committed:
update test.ts

Pranavchiku committed:
apply suggestion from code review in assign.js

Pranavchiku commented:
I updated typescript tests, applied recent code review, all tests and benchmarks cleared, this PR can have a review.

Pranavchiku committed:
remove blank line

Pranavchiku committed:
wrap all floating point operations with float64ToFloat32

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,128 @@
+	srotg( 0.0, '5' ); // $ExpectError
+	srotg( 0.0, [] ); // $ExpectError
+	srotg( 0.0, {} ); // $ExpectError
+	srotg( true, 2.0 ); // $ExpectError
```
```suggestion
	srotg( true, 2.0 ); // $ExpectError

```

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,128 @@
+{
+	const out = new Float32Array( 4 );
+
+	srotg( 0.0, true, out, 1, 0 ); // $ExpectError
+	srotg( 0.0, false, out, 1, 0 ); // $ExpectError
+	srotg( 0.0, null, out, 1, 0 ); // $ExpectError
+	srotg( 0.0, undefined, out, 1, 0 ); // $ExpectError
+	srotg( 0.0, '5', out, 1, 0 ); // $ExpectError
+	srotg( 0.0, [], out, 1, 0 ); // $ExpectError
+	srotg( 0.0, {}, out, 1, 0 ); // $ExpectError
+	srotg( true, 2.0, out, 1, 0 ); // $ExpectError
+	srotg( false, 2.0, out, 1, 0 ); // $ExpectError
+	srotg( null, 2.0, out, 1, 0 ); // $ExpectError
+	srotg( undefined, 2.0, out, 1, 0 ); // $ExpectError
+	srotg( '5', 2.0, out, 1, 0 ); // $ExpectError
+	srotg( [], 2.0, out, 1, 0 ); // $ExpectError
+	srotg( {}, 2.0, out, 1, 0 ); // $ExpectError
```
```suggestion
	srotg.assign( 0.0, true, out, 1, 0 ); // $ExpectError
	srotg.assign( 0.0, false, out, 1, 0 ); // $ExpectError
	srotg.assign( 0.0, null, out, 1, 0 ); // $ExpectError
	srotg.assign( 0.0, undefined, out, 1, 0 ); // $ExpectError
	srotg.assign( 0.0, '5', out, 1, 0 ); // $ExpectError
	srotg.assign( 0.0, [], out, 1, 0 ); // $ExpectError
	srotg.assign( 0.0, {}, out, 1, 0 ); // $ExpectError

	srotg.assign( true, 2.0, out, 1, 0 ); // $ExpectError
	srotg.assign( false, 2.0, out, 1, 0 ); // $ExpectError
	srotg.assign( null, 2.0, out, 1, 0 ); // $ExpectError
	srotg.assign( undefined, 2.0, out, 1, 0 ); // $ExpectError
	srotg.assign( '5', 2.0, out, 1, 0 ); // $ExpectError
	srotg.assign( [], 2.0, out, 1, 0 ); // $ExpectError
	srotg.assign( {}, 2.0, out, 1, 0 ); // $ExpectError
```

kgryte committed:
Apply suggestions from code review

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/docs/types/test.ts:
```diff
@@ -0,0 +1,130 @@
+	srotg( 0.0, '5' ); // $ExpectError
+	srotg( 0.0, [] ); // $ExpectError
+	srotg( 0.0, {} ); // $ExpectError
+	srotg( true, 2.0 ); // $ExpectError
+
```
```suggestion

	srotg( true, 2.0 ); // $ExpectError
```

kgryte committed:
Fix line break

kgryte commented on lib/node_modules/@stdlib/blas/base/srotg/lib/assign.js:
```diff
@@ -0,0 +1,95 @@
+		r = 0.0;
+		z = 0.0;
+	} else {
+		r = float64ToFloat32( scale * float64ToFloat32( sqrtf( pow( float64ToFloat32( a/scale ), 2 ) + pow( float64ToFloat32( b/scale ), 2 ) ) ) ); // eslint-disable-line max-len
```
```suggestion
		r = float64ToFloat32( scale * float64ToFloat32( sqrtf( float64ToFloat32( pow( float64ToFloat32( a/scale ), 2 ) + pow( float64ToFloat32( b/scale ), 2 ) ) ) ) ); // eslint-disable-line max-len
```

kgryte committed:
fix: add missing cast

kgryte reviewed (approved):
LGTM. Thanks, @Pranavchiku!
